File: src/kotlin/src/main/kotlin/AlgebraicDataTypes.kt
//: src/kotlin/src/main/kotlin/AlgebraicDataTypes.kt
package algebraic
import algebraic.ADT.*

sealed class ADT {
    class Person(val eats: String) : ADT()
    class Robot(val chargesWith: String): ADT()
}

fun nourish(subject: ADT) {
    when (subject) {  // Exhaustive
        is Person -> println("eating ${subject.eats}")
        is Robot -> println("charging with ${subject.chargesWith}")
    }
}

fun main() = listOf(Person("pizza"), Robot("electricity")).forEach { nourish(it) }

File: src/scala/AlgebraicDataTypes.scala
//: src/scala/AlgebraicDataTypes.scala
package adts
import ADT.*

enum ADT:
    case Person(val eats: String)
    case Robot(val chargesWith: String)

def nourish(x: ADT) = x match   // Exhaustive
    case p: Person => println(s"eating ${p.eats}")
    case r: Robot  => println(s"charging with ${r.chargesWith}")

@main def main() =
    List(Person("pizza"), Robot("electricity")).foreach(nourish)
